rosbag2_transport/src/rosbag2_transport/player.cpp

```diff
@@ -328,6 +350,134 @@ bool Player::play_next()
+  }
+  if (current_message->time_stamp != origin_message->time_stamp) {
+    throw std::runtime_error("Fail to restore message queue");
+  }
+  message_queue_.enqueue(current_message);
+  enqueue_up_to_boundary(play_options_.read_ahead_queue_size);  // refill queue
+}
+
+bool Player::seek(rcutils_time_point_value_t time_point)
+{
+  // Temporary stop playback in play_messages_from_queue() and block play_next()
+  std::lock_guard<std::mutex> main_play_loop_lk(skip_message_in_main_play_loop_mutex_);
+  skip_message_in_main_play_loop_ = true;
+  // Wait for player to be ready for playback messages from queue i.e. wait for Player:play() to
+  // be called if not yet and queue to be filled with messages.
+  {
+    std::unique_lock<std::mutex> lk(ready_to_play_from_queue_mutex_);
+    ready_to_play_from_queue_cv_.wait(lk, [this] {return is_ready_to_play_from_queue_;});
+  }
+
+  cancel_wait_for_next_message_ = true;
+  bool message_found = false;
+  rosbag2_storage::SerializedBagMessageSharedPtr * message_ptr = peek_next_message_from_queue();
+  rosbag2_storage::SerializedBagMessageSharedPtr message_before_seek;
+  if (message_ptr != nullptr) {
+    message_before_seek = *message_ptr;
+  }
+
+  {  // Update message queue and adjust reader pointer
+    std::lock_guard<std::mutex> lk(reader_mutex_);
+    // Read next current_message
+    rosbag2_storage::SerializedBagMessageSharedPtr current_message;
+    if (!message_queue_.try_dequeue(current_message)) {
+      if (!reader_->has_next()) {
+        reader_->close();
+        reader_->open(storage_options_, {"", rmw_get_serialization_format()});
+        if (!reader_->has_next()) {
+          return false;  // The bag is empty
+        }
+        // else means that we were at the end of the bag when we were started and will try to
+        // search from the beginning
+        assert(is_storage_completely_loaded());
+        storage_loading_future_ =
+          std::async(std::launch::async, [this]() {load_storage_content();});
+      }
+      current_message = reader_->read_next();
+    }
+
+    const bool jump_forward = (time_point >= current_message->time_stamp);
+
+    if (jump_forward) {
+      if (current_message->time_stamp >= time_point) {
+        message_found = true;
+      }
+      // Pop up current_message queue trying to find current_message with needed timestamp
+      while (!message_found && message_queue_.try_dequeue(current_message)) {
+        if (current_message->time_stamp >= time_point) {
+          message_found = true;
+        }
+      }
+      while (!message_found && reader_->has_next()) {
+        current_message = reader_->read_next();
+        if (current_message->time_stamp >= time_point) {
+          message_found = true;
+        }
+      }
+      if (message_found) {
+        // Enqueue current_message and rest of the queue to the queue again to preserve order of
+        // messages
+        if (message_queue_.peek() != nullptr) {  // if message_queue_ not empty
```

Comment:
nit: it might be nice to pull this logic into a separate function so it's easier to read - like `empty_queue_into(from, into)`.

Reply:
I would like to keep it as is. wrapped in function it will be unclear what it does why.  
It's just a few lines of code in place. I can't move out whole block of code since current message is not empty already and should be accounted.
```cpp
         decltype(message_queue_) temp_message_queue;
          temp_message_queue.enqueue(current_message);
          // Copy residual messages from message_queue_ to the temp_message_queue
          while (message_queue_.try_dequeue(current_message)) {
            temp_message_queue.enqueue(current_message);
          }
```
It would be possible to move out only 
```cpp
          while (message_queue_.try_dequeue(current_message)) {
            temp_message_queue.enqueue(current_message);
          }
```
But it's really doesn't make sense to do this IMO.